### ISSUE TITLE
Pre-built firmware?

### ISSUE THREAD

**Issue by johncblacker** [rcludwick/Lora_driver_for_MicroPython_on_ttgo#13]
I've been trying to build the ESP32 version of MP to no avail.  Any possibility that you have a pre-built binary for the TTGO board you could make available?
Thanks.
John Blacker

**Comment by rcludwick** [comment]
I don't have one handy, though I've built one in the past.  It will take me
some time to get to it.

I was contemplating building a docker image that does this.  Would you be
interested in this?

**Comment by johncblacker** [comment]
Hey, anything would help, although I found the "lobo" fork and it has binaries which seem to work.  Anyway, go for it!
Thanks.

**Comment by johncblacker** [comment]
How do I accomplish step 2 (link ...... into MP esp32 directory)?  Not sure exactly what you mean.

**Comment by johncblacker** [comment]
trying to use the wei1234C repository driver, but keep getting invalid version .  I used an arduino program to dump the registers of my board and at 42 I have 0X12 which is what the driver is supposed to be looking for; however, when it runs, it picks up 0xff (255) instead and won't go on.  Don't know what could be wrong and attempts to get info on wei1234C go unanswered.

**Comment by rcludwick** [comment]
link means 'ln -s'

Basically check out this source and the micropython source and then use 'ln
-s' to create a symbolic link from this library to a directory under the
esp32 micropython source directory.

Yeah, as to wei1234C, the guy did some good work getting it going, but the
code quality was rather lacking.  When I suggested trying to submit a pull
request he insisted that it be tested on all the platforms it supported.
But.. I only have the TTGO, so that was a no go.  Hence the fork.

This project needs some love, and I don't have a lot of time to do it
unfortunately.  But I'll see if I can carve out some time to build esp32
binaries again.

**Comment by johncblacker** [comment]
Hold off a bit, I've finally gotten over the problems building MP for ESP32 and so I'll insert the links and give it a go and see what comes out.  Will report back.

**Comment by johncblacker** [comment]
still a little unsure about the links.  Should I create shadow directory in the modules directory and put the links inside that?

**Comment by johncblacker** [comment]
Rather than rebuild with the links, I just copied the appropriate directories over to the board, put the appropriate directories into sys.path and did an import of test.py.  Now, what if any output should I see?   I'm running in a terminal.

**Comment by rcludwick** [comment]
There should be output on the screen.

https://photos.app.goo.gl/6Aooe9RRLhkFDkUg9

**Comment by johncblacker** [comment]
OK, I see now, I've setup another ttgo, but I'm not seeing anything from either one of them.  To get it going, I've just done a import test, don't know if that'll do it or not.  Relative to setting up the MP repository with links, could you describe the directory structure you have within the modules subdirectory?  Thanks.

**Comment by johncblacker** [comment]
Got it working, although I had to comment out the version check in sx127x.py.  I have one bad display, so I'm going to try a different one - I have an I2c and an SPI version.  Theoretically, the SPI version should work being daisy chained off the SPI that the sx127x is using, although there might be some timing issues with the sx127x - we'll see.

**Comment by johncblacker** [comment]
Running 2 instances of test on two different TTGO's, but they both seem to "think" they're senders?  One's sitting at the "hello !" message, the other is plugging away sending and displaying info.  What am I doing wrong?

**Comment by rcludwick** [comment]
I remember that one needs to run as a sender and another as a receiver.
I'll look into it on the plane today.

**Comment by rcludwick** [comment]
The duplex demo is supposed to do both send and receive.

**Comment by johncblacker** [comment]
To clarify...I'm seeing output on one old that looks normal and the REPL is giving me the sending stuff; on the other oled, I'm just seeing "Hello !" but the REPL is giving me normal sending output stuff.

**Comment by johncblacker** [comment]
FYI I was successful in converting one of my boards to output to a ST7735 using spi.  I ended up using some now defunct (but useful, nonetheless) Adafruit modules from the early days of CircuitPython (rgb.py, gfx.py, bitmapfont.py, font5x8.bin, and st7735) and created another controller_esp_ttgo_lora_spi.py module and modified the config.py module adding another variable: IS_TTGO_LORA_SPI.

However, I still have issues where both boards are trying to send only, they never get to receive (I don't see any "print" output in REPL and I don't see the displays change to output the RSSI info, they both display "Hello !" and I've tried restarting them randomly to see if it was some kind of timing issue to no avail.  Any ideas?

**Comment by rcludwick** [comment]
So I won't be able to get back to this until this weekend.  As I recall the
chip stores the message until it's asked for.  So it shouldn't matter when
one receives and one transmits.

Maybe I can set up a time we can chat through this.  Possibly Saturday?  If
not then maybe Friday?

**Comment by johncblacker** [comment]
Thanks for getting back to me.  I don’t know what happened, but it started working like it’s supposed to…each board doing alternating send/receives.  I don’t think it 

was anything I did.  Anyway, I’m satisfied with the outcome.  I now have both the oled and spi displays working, so I’m happy.  The issue I have with MP is that they’re 

changing the API’s and there’s so many versions of doc out there it’s sometimes difficult to figure out what version to use.  Also, each port can differ for things that should be

very similar – like SPI.  I found different different signatures on spi in ESP8266/32 and other MP spi impllementations.  

 

One of the things I like about CircuitPython is that they make things quite simple; whereas the same thing in MP is more complex.  I just got two RFM9X breakout boards from

Adafruit and in less than 30 minutes was exchanging messages between boards…now that’s the way it should be.  The main program was less 20 lines of code and the “helper”

code is 646 lines including comments and is easy to follow.  Going to look around for a CP port for the sx127x; but, your version is now working for me.  I went back and looked at

Wei’s version and just shook my head and gave up on it.  

 

What I try to do when trying to implement something in either MP or CP is find some Arduino code and simply convert that over, without trying to overcomplicate theprocess.

You have two files to convert: .cpp and .h and it’s not that hard; but, I do always look for an already converted implementation and give that a try.

 

Anyway, thanks for your good work and for being patient with me throughout this past week.  I appreciate your help immensely!

jb
